### DISCUSSION TITLE
Downtimes with volatile packet is discarded since the transport is not currently writable

### DISCUSSION THREAD

**Discussion by severinolorillajr** [socketio/socket.io#5063] [Q&A]
Our socket server generally works fine.

    "socket.io": "^4.5.3",
    "socket.io-client": "^4.5.4",

But we've been having downtimes twice a week (more or less).
We've been bombarded with these messages:

Jul  9 02:09:36 ip-10-0-31-159 web[255737]: 2024-07-09T02:09:36.802Z socket.io:client volatile packet is discarded since the transport is not currently writable
Jul  9 02:09:36 ip-10-0-31-159 web[255737]: 2024-07-09T02:09:36.802Z socket.io:client volatile packet is discarded since the transport is not currently writable
Jul  9 02:09:36 ip-10-0-31-159 web[255737]: 2024-07-09T02:09:36.802Z socket.io:client volatile packet is discarded since the transport is not currently writable
Jul  9 02:09:36 ip-10-0-31-159 web[255737]: 2024-07-09T02:09:36.802Z socket.io:client volatile packet is discarded since the transport is not currently writable
Jul  9 02:09:36 ip-10-0-31-159 web[255737]: 2024-07-09T02:09:36.804Z socket.io:client volatile packet is discarded since the transport is not currently writable
Jul  9 02:09:36 ip-10-0-31-159 web[255737]: 2024-07-09T02:09:36.805Z socket.io:client volatile packet is discarded since the transport is not currently writable

Any ideas how to approach or work around this?
Can't seem to find anything on Socket IO docs and the web discussing this.

Was this because we are looping volatile emit events?
But it doesn't seem to happen that frequently though.

**Comment by darrachequesne** [comment]
Hi! What do you mean by "downtimes"? Does your server crash?

Volatile packets can indeed be discarded if the underlying transport is not ready to be used, for example if there is no pending GET request when using HTTP long-polling.

Reference: https://socket.io/docs/v4/emitting-events/#volatile-events

**Reply by severinolorillajr** [reply]
Servers are still up as I still see logs coming out.
It was flooded by the discarded packets error above though.
All clients are using the websocket transport.

What could be the cause of transport not ready or not writeable? 
does the volume or traffic matter?
Don't seem to have issues around 95% of the time.

**Reply by darrachequesne** [reply]
> What could be the cause of transport not ready or not writeable?

The WebSocket transport can become not writable when a lot of data gets written. In that case, packets tagged as "volatile" can indeed be discarded. If you don't need it, you can simply remove the `.volatile` modifier:

```diff
- socket.volatile.emit("my-event");
+ socket.emit("my-event");
```

**Reply by severinolorillajr** [reply]
I want to use volatile since I only need the current (latest) state of the app.
Is that taken into account as whole? Like if I have volatile events for different destinations, will that matter?
My problem here though once I reach a state where I'm spammed with these errors.
I usually need a server reboot to get it back up and running.

Would there be a way for us to keep track when a transports is about to reach it's limit?
And how do I extend it so the errors can be minimized?

**Reply by darrachequesne** [reply]
> Like if I have volatile events for different destinations, will that matter?

The server checks for the state of each client individually. The volatile packets may be discarded for some, and sent for others.

> My problem here though once I reach a state where I'm spammed with these errors.

Please note that these are debug logs, not errors. This might suggest an actual problem though, as rebooting the server should not be needed.

> Would there be a way for us to keep track when a transports is about to reach it's limit?

I guess you could check the `bufferedAmount` value of the WebSocket object, though I'm not sure how it can help in your case.

Reference: https://github.com/websockets/ws/blob/master/doc/ws.md#websocketbufferedamount

**Reply by severinolorillajr** [reply]
Appreciate the time you spent answering my questions @darrachequesne.
I think this is an issue on code level from our side.

I'll add more once I have more information.
For now, I am closing this discussion.

Cheers!

**Reply by severinolorillajr** [reply]
@darrachequesne just a quick confirmation though.
You mentioned that websocket transport can be come unwritable when a lot of data gets written.
And in return could result to dropped packets.

Have a few questions though:
1. If I added a "polling" transport would it automatically downgrade from websocket when transport becomes unwriteable?
2. If 1 above automatically does that, will the websocket transport be able to recover, i.e. become writable again?
3. If I am downgraded to polling, is there a mechanism that automatically reverts it back to websocket once available?
4. If i want to have polling as fallback.. i can add the transports in the clients in this order right? websocket, then polling?

**Reply by severinolorillajr** [reply]
Add'l info here:

I noticed that the dropped packets has been going on for sometime.
Even if we did not really experience the server being unresponsive.
But the volume of this dropped packets, is already concerning.

And at times, it makes the system totally unusable, needing a reboot.

**Reply by darrachequesne** [reply]
The transport becomes unwritable for a short amount of time (encoding + writing to underlying WebSocket): https://github.com/socketio/socket.io/blob/b79d80aa5989b21c22f4901a2d1b53cf51351020/packages/engine.io-client/lib/transports/websocket.ts#L91-L120

There is currently no way to downgrade: by default, polling is used first, and upgraded to WebSocket if possible.

There will be a few performance improvements on the server side in the next version, maybe that will help ([diff](https://github.com/socketio/engine.io/compare/6.5.4...6.6.0)).

**Reply by severinolorillajr** [reply]
Reopening this discussion again.

We reached a point where we are bottlednecked at socket.io level again.
<img width="1671" height="193" alt="image" src="https://github.com/user-attachments/assets/17242043-a91a-47bc-9305-c349832c1ae8" />
The discarded packets above is quite alarming.
This happened as our user base grow.

By the way, we are already using a cluster-worker setup via redis adapters.
Currently we have 8 workers up and running.
We are eventually planning to transitition to message queues with socket.io.
But we are concerned that even if we actively manage the volume of messages, we will still be blocked by socket.io.

Any recommendations?

**Reply by severinolorillajr** [reply]
Have already exhausted all available information from socket.io's documentation.
https://socket.io/docs/v4/performance-tuning/
But did not give us any workable solution.

**Reply by juanjaho** [reply]
It sounds like you're hitting some fundamental scaling challenges with Socket.IO's underlying transport layer.

To help narrow down the cause, could you clarify a few things?
-  Packet Dropping: Are these discarded packets occurring when you use `socket.emit` or `socket.volatile.emit`? `socket.volatile.emit` is designed to intentionally drop packets, whereas `socket.emit` is meant to be reliable and should not be dropping packets without a deeper issue. If you are using `socket.volatile.emit`, one potention cause is simply that sender is sending the message to early and suffer from the following https://github.com/socketio/socket.io/issues/3350

-  Server Crashes: If you're using `socket.emit`, are you also experiencing any server crashes? If so, have you observed a pattern of increasing RAM usage leading up to these crashes? This could be a sign that the server's buffer is building up faster than it can send data, eventually exhausting the allocated memory. If so you're likely facing is related to the "head-of-line blocking" problem at both transport layer and application layer inherent in the TCP protocol that WebSocket and HTTP/2 rely on. Even with a cluster-worker setup and Redis adapters, a single slow consumer on a single TCP connection can slow down the entire stream for all other messages and clients on that **same connection**.

Given your plan to move to a message queue, you're on the right track for actively managing message volume and backpressure. However, this might not fully solve the bottleneck at the transport layer.
A strong recommendation for a long-term solution would be to explore upgrading to HTTP/3 and the WebTransport protocol with Socket.IO which doesn't suffer from HoL at transport layer blocking that TCP does. And it's simply a faster protocol.

As for application layer HoL issues, it will have to be fixed with a better message parsing architecture at code level.

**Reply by severinolorillajr** [reply]
Thank you for alloting time to answer my queries and providing recommendations.
This is greatly appreciated @juanjaho.

- Packet dropping: we actually have a mix of both volatile and regular emit on our use cases. Don't have an actual data on the ratios on the volume per sec. Our volatile emits are usually for game updates per player, so it could easily spike as the number of users grow,

- Server crashes: as of late, we are no longer experiencing the crashes. But the discarded packets remained constant, and kept increasing as our user base grows. Our instance sizes are generally oversized for it's usage, we are only using 5-20% CPU and memory.

**Reply by juanjaho** [reply]
A consistent ratio of dropped volatile emit packets to total volatile emit packets suggests a normal, expected rate of packet loss. If this ratio remains constant, for example, 10% of packets are dropped, then an increase in total packets sent will result in a proportional increase in dropped packets. This is not necessarily a cause for concern.

However, if the ratio of dropped packets to sent packets begins to increase, it indicates a growing problem with packet loss. For instance, if the ratio jumps from 10% to 20%, it suggests that the system's ability to handle the load is deteriorating. This is when you should investigate the cause of the increased packet loss.

Do you have data points to analyse the ratio of packet loss?
